Create deb repository root

ycointe opened the pull request:
**Component**: Buildchain

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: MetalK8son Ubuntu

**Summary**: We need to create Debian repository root directly in the buildchain and not in the script which download packages.

**Acceptance criteria**: Execute task download_deb_packages with the command: `./doit.sh _download_deb_packages` the verify in the folder ` _build/root/packages/` that you have all repository folder with .deb corresponding inside


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1701

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ycointe committed:
build/packages: Add the list of used Debian repositories

We add the different Debian repositories from the packages are downloaded to
reproduce the same architecture in the futur internal repository.

Refs: #1701

ycointe committed:
build/packages: Create the folders for downloaded packages

We want to have downloaded packages in some folder which are named according to
the repository name from there are downloaded.
We create these folder with the list of Debian repositories that we
created before.

Refs: #1701

ycointe committed:
build/package: Remove the folder creation

We remove the folder creation in the script. This is handled directly by
the buildchain.

Refs: #1701

bert-e commented:
# Hello ycointe,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

slaperche-scality commented on packages/debian/Dockerfile:
```diff
@@ -12,7 +12,7 @@ RUN apt update -y \
         lintian \
         apt-rdepends \
-        python-apt \
+        reprepro \
```
Not related to this PR

ycointe replied:
Indeed, I cleaned up.

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -404,6 +404,48 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
         if package.name not in DEB_TO_BUILD
+    )
+
+SCALITY_DEB_REPOSITORY = targets.DEBRepository(
```
Why not declare it directly in `DEB_REPOSITORIES` like the others?

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -153,6 +153,10 @@ def clean() -> None:
 def task__download_deb_packages() -> types.TaskDict:
     """Download Debian packages locally."""
     # TODO: Clean the repository
```
Now that we `mkdirs` we could have a commit implementing the cleanup.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/1701-create-deb-repository-root` and the
destination branch `development/2.4`.

Please resolve the conflict on **the feature branch** (`improvement/1701-create-deb-repository-root`).

```bash
 $ git fetch
 $ git checkout origin/improvement/1701-create-deb-repository-root
 $ git merge origin/development/2.4
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/1701-create-deb-repository-root
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -175,7 +175,21 @@ def clean() -> None:
-    # TODO: Clean the repository
+    def clean() -> None:
+        """Delete cache and repositories on the ISO."""
+        coreutils.rm_rf(constants.REPO_DEB_ROOT)
```
If you `rm -rf` that directory, you also remove all the repo at once (so the loop is useless).
I think we want something like that instead:

```python
        for repository in DEB_REPOSITORIES:
            # Repository with an explicit list of packages are created by a
            # dedicated task that will also handle their cleaning, so we skip
            # them here.
            if repository.packages:
                continue
            coreutils.rm_rf(repository.rootdir)
        constants.REPO_DEB_ROOT.rmdir()
```

ycointe committed:
build/packages: Add class DEBRepository

We need the class DEBRepository to create Debian repository.
For now the class is only use to permit the Debian packages download.

Refs: #1701

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

ycointe commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/1bd91c224414692e76b82f7c57fd27e33d3cc69e) did not succeed in branch improvement/1701-create-deb-repository-root.




*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ycointe.
